[AArch64][GlobalISel] More FCmp legalization.

davemgreen opened the pull request:
This fills out the fcmp handling to be more like the other instructions, adding better support for fp16 and some larger vectors.

Select of f16 values is still not handled optimally in places as the select is only legal for s32 values, not s16. This would be correct for integer but not necessarily for fp. It is as if we need to do legalization -> regbankselect -> extra legaliation -> selection.

There is some fallback due to some missing handling for shifts, which I think Chuong is currently looking at.

arsenm commented on llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp:
```diff
@@ -1556,6 +1556,16 @@ LegalizerHelper::LegalizeResult LegalizerHelper::narrowScalar(MachineInstr &MI,
+    if (TypeIdx == 0)
+      narrowScalarDst(MI, NarrowTy, 0, TargetOpcode::G_ZEXT);
+    else {
+      return UnableToLegalize;
```
Turn this into early return, you are calling the observer on the failure path

tschuett commented on llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp:
```diff
@@ -566,8 +553,10 @@ AArch64LegalizerInfo::AArch64LegalizerInfo(const AArch64Subtarget &ST)
+      .clampNumElements(1, v4s16, v8s16)
+      .clampNumElements(1, v2s32, v4s32)
+      .clampMaxNumElements(1, s64, 2)
+      .moreElementsToNextPow2(1);
```
`moreElementsVector` does not exist for G_FCMP:
https://github.com/llvm/llvm-project/blob/d0d072710468316edbd4130e50f1146c5a6aca89/llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp#L5317

davemgreen replied:
Yeah I had added it, but it was still falling back due to the unsupported shifts. Now that those are in it's looking better.

tschuett commented on llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp:
```diff
@@ -532,32 +532,19 @@ AArch64LegalizerInfo::AArch64LegalizerInfo(const AArch64Subtarget &ST)
+                 {v2s64, v2s64}})
+      .legalIf([=](const LegalityQuery &Query) {
+        const auto &Ty = Query.Types[0];
+        return (Ty == v8s16 || Ty == v4s16) && HasFP16;
```
It reads as if arbitrary types on the RHS are legal.

tschuett commented on llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp:
```diff
@@ -566,8 +553,10 @@ AArch64LegalizerInfo::AArch64LegalizerInfo(const AArch64Subtarget &ST)
-      .clampMaxNumElements(1, s64, 2);
+      .clampNumElements(1, v4s16, v8s16)
+      .clampNumElements(1, v2s32, v4s32)
+      .clampMaxNumElements(1, s64, 2)
```
```
.clampNumElements(1, v2s64, v2s64)
```

davemgreen replied:
There is only a max amount here, as it can't be a v1 vector. You could argue that keeping them more consistent is more readable, but only using clampNumElements is less legalizer rules.

tschuett commented:
G_SELECT shows an optimization opportunity, if the LLT could distinguish between i16 and f16. i16 would never be legal and f16 would sometimes be legal for G_SELECT. The costs are that we have to legalize G_SELECT for i64 and f64.

davemgreen reviewed (commented):
> G_SELECT shows an optimization opportunity, if the LLT could distinguish between i16 and f16. i16 would never be legal and f16 would sometimes be legal for G_SELECT. The costs are that we have to legalize G_SELECT for i64 and f64.

Yeah that's what I was saying about "It is as if we need to do legalization -> regbankselect -> extra legaliation -> selection", at least in the current design without fp types.

aemerson reviewed (approved):
LGTM, thanks!

davemgreen committed:
[AArch64][GlobalISel] More FCmp legalization.

This fills out the fcmp handling to be more like the other instructions, adding
better support for fp16 and some larger vectors.

Select of f16 values is still not handled optimally in places as the select is
only legal for s32 values, not s16. This would be correct for integer but not
necessarily for fp. It is as if we need to do legalization -> regbankselect ->
extra legaliation -> selection.

There is some fallback due to some missing handling for shifts, which I think
Chuong is currently looking at.
